src/Features/CSharp/Portable/UseRecursivePatterns/CSharpUseRecursivePatternsCodeRefactoringProvider.AnalyzedNode.cs

```diff
@@ -0,0 +1,385 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseRecursivePatterns
+{
+    internal sealed partial class CSharpUseRecursivePatternsCodeRefactoringProvider
+    {
+        private abstract class AnalyzedNode : Visitor<AnalyzedNode>
+        {
+            public abstract T Accept<T>(Visitor<T> visitor);
+            public abstract override string ToString();
+
+            public sealed override AnalyzedNode VisitSourcePattern(SourcePattern node) => throw ExceptionUtilities.Unreachable;
+        }
+
+        private sealed class Conjuction : AnalyzedNode
+        {
+            public readonly AnalyzedNode Left;
+            public readonly AnalyzedNode Right;
+
+            public Conjuction(AnalyzedNode left, AnalyzedNode right) => (Left, Right) = (left, right);
+            public override T Accept<T>(Visitor<T> visitor) => visitor.VisitConjuction(this);
+            public override string ToString() => $"{Left} AND {Right}";
+
+            public override AnalyzedNode VisitConjuction(Conjuction node)
+            {
+                throw new NotImplementedException();
+            }
+
+            public override AnalyzedNode VisitConstantPattern(ConstantPattern node)
+            {
+                throw new NotImplementedException();
+            }
+
+            public override AnalyzedNode VisitNotNullPattern(NotNullPattern node)
+            {
+                throw new NotImplementedException();
+            }
+
+            private sealed class Finder : Visitor<bool>
+            {
+                private readonly ExpressionSyntax _expression;
+
+                public Finder(ExpressionSyntax expression) => _expression = expression;
+                public override bool VisitConjuction(Conjuction node) => Visit(node.Left) || Visit(node.Right);
+                public override bool VisitConstantPattern(ConstantPattern node) => false;
+                public override bool VisitNotNullPattern(NotNullPattern node) => false;
+                public override bool VisitPatternMatch(PatternMatch node) => SyntaxFactory.AreEquivalent(_expression, node.Expression);
+                public override bool VisitSourcePattern(SourcePattern node) => false;
+                public override bool VisitTypePattern(TypePattern node) => false;
+                public override bool VisitVarPattern(VarPattern node) => false;
```

Comment:
these all need explanations.  for example, why _expression could not be found in a constant pattern

Reply:
because the expression we're trying to find would not possibly appear in a constant pattern. will add a comment.